pkg/storage/storage.go

```diff
@@ -42,4 +43,7 @@ type ResourceStorageConfig struct {
 	Codec          runtime.Codec
 	StorageVersion schema.GroupVersion
 	MemoryVersion  schema.GroupVersion
+	WatchCache     *watchcache.WatchCache
```

Comment:
An interface can be defined to avoid dependencies on the implementation.

Reply:
Do you mean add a new interface in ResourceStorage like this ?
```
type ResourceStorage interface {
	GetStorageConfig() *ResourceStorageConfig

	GetWatchCache() *watchcache.WatchCache

	Get(ctx context.Context, cluster, namespace, name string, obj runtime.Object) error
	List(ctx context.Context, listObj runtime.Object, opts *internal.ListOptions) error

	Create(ctx context.Context, cluster string, obj runtime.Object) error
	Update(ctx context.Context, cluster string, obj runtime.Object) error
	Delete(ctx context.Context, cluster string, obj runtime.Object) error
}

```

Reply:
I'm looking at the memory storage implementation, and config as a static structure really shouldn't be used to feed information.

```go
type ResourceStorage interface {
	GetStorageConfig() *ResourceStorageConfig

	GetWatchCache() *watchcache.WatchCache

	Get(ctx context.Context, cluster, namespace, name string, obj runtime.Object) error
	List(ctx context.Context, listObj runtime.Object, opts *internal.ListOptions) error

	Create(ctx context.Context, cluster string, obj runtime.Object) error
	Update(ctx context.Context, cluster string, obj runtime.Object) error
	Delete(ctx context.Context, cluster string, obj runtime.Object) error
}
```
This looks better, and it is recommended not to rely on `*watchcache.WatchCache`, but to abstract an interface such as Watcher

Reply:
```
type ResourceStorage interface {
	GetStorageConfig() *ResourceStorageConfig

	GetWatchCache() *watchcache.WatchCache

	Get(ctx context.Context, cluster, namespace, name string, obj runtime.Object) error
	List(ctx context.Context, listObj runtime.Object, opts *internal.ListOptions) error

	Create(ctx context.Context, cluster string, obj runtime.Object) error
	Update(ctx context.Context, cluster string, obj runtime.Object) error
	Delete(ctx context.Context, cluster string, obj runtime.Object) error
}
```
`GetWatchCache() ` is only for memory storage, is that ok for other storages?


Or we just do like this, Watcher is an interface.
```
type ResourceStorageConfig struct {
	GroupResource        schema.GroupResource
	StorageGroupResource schema.GroupResource

	Codec          runtime.Codec
	StorageVersion schema.GroupVersion
	MemoryVersion  schema.GroupVersion
	WatchCache     Watcher
	Cluster        string
	Namespaced     bool
}
```

Reply:
Don't let the caller get additional information from the configuration passed to the function again.

> GetWatchCache() is only for memory storage, is that ok for other storages?

Other storage layers may also implement watch functions, `GetWatchCache`/`GetWatcher`/`Watcher` should be a general method of the Resource Storage interface, whether or not to implement it is up to the specific storage layer.
```go
type ResourceStorage interface {
	GetStorageConfig() *ResourceStorageConfig

	GetWatchCache() ResourceWatcher
```

Reply:
Got it~

Reply:
How about just add watch func, like get and list?
```
type ResourceStorage interface {
	GetStorageConfig() *ResourceStorageConfig

	Get(ctx context.Context, cluster, namespace, name string, obj runtime.Object) error
	List(ctx context.Context, listObj runtime.Object, opts *internal.ListOptions) error
	Watch(ctx context.Context, options *metainternalversion.ListOptions) (watch.Interface, error)

	Create(ctx context.Context, cluster string, obj runtime.Object) error
	Update(ctx context.Context, cluster string, obj runtime.Object) error
	Delete(ctx context.Context, cluster string, obj runtime.Object) error
}
```

Reply:
This way looks better